tests/integration/conftest.py

```diff
@@ -0,0 +1,112 @@
+import logging
+import pathlib
+import socket
+import tempfile
+import time
+
+import pytest
+import trio
+
+
+@pytest.fixture
+def dopple_file():
+    return pathlib.Path(__file__).parent.parent.parent / "dopple" / "dopple.py"
+
+
+def generate_docker_cmds(name, cmd_continuation):
+    """
+    Generate ``start`` and ``stop`` commands for any given cmd that is intended to use
+    docker under the current host's user account.
+    """
+    start_cmd = (
+        f"docker run --rm --name {name} " f"--user $(id -u):$(id -g) {cmd_continuation}"
+    )
+
+    stop_cmd = f"docker stop {name}"
+
+    return start_cmd, stop_cmd
+
+
+@pytest.fixture
+async def aleth_ipc_path():
+    """
+    Run aleth inside docker using a shared temporary directory on the host,
+    then yield its IPC path.
+    """
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmpdir_name = pathlib.Path(tmpdir).name
+
+        start_cmd, stop_cmd = generate_docker_cmds(
+            tmpdir_name, f"-v {tmpdir}:/.ethereum ethereum/aleth:1.6.0"
+        )
+
+        ipc_path = f"{tmpdir}/geth.ipc"
+        logging.debug(f"Starting aleth, ipc path: {ipc_path}")
+        async with await trio.open_process(start_cmd, shell=True):
+            wait_for_socket(ipc_path)
+            yield ipc_path
+            logging.debug("Killing aleth")
+            await trio.run_process(stop_cmd, shell=True)
+
+
+@pytest.fixture
+async def geth_ipc_path():
+    """
+    Run geth inside docker using a shared temporary directory on the host,
+    then yield its IPC path.
+    """
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmpdir_name = pathlib.Path(tmpdir).name
+
+        start_cmd, stop_cmd = generate_docker_cmds(
+            tmpdir_name,
+            f"-v {tmpdir}:/tmp/.ethereum ethereum/client-go:v1.9.6 --datadir /tmp/.ethereum",
+        )
+
+        ipc_path = f"{tmpdir}/geth.ipc"
+        logging.debug(f"Starting geth, ipc path: {ipc_path}")
+        async with await trio.open_process(start_cmd, shell=True):
+            wait_for_socket(ipc_path)
+            yield ipc_path
+            logging.debug("Killing geth")
+            await trio.run_process(stop_cmd, shell=True)
+
+
+@pytest.fixture
+async def dopple_as_file_on_geth(geth_ipc_path, dopple_file):
+    """
+    Run geth, then run dopple as a script file to connect to it.
+    """
+    async with await trio.open_process([dopple_file, geth_ipc_path]) as proc:
+        await trio.sleep(0.5)
+        yield
+        proc.terminate()
+        await proc.wait()
+
+
+@pytest.fixture
+async def dopple_as_file_on_aleth(aleth_ipc_path, dopple_file):
+    """
+    Run aleth, then run dopple as a script file to connect to it.
+    """
+    async with await trio.open_process([dopple_file, aleth_ipc_path]) as proc:
```

Comment:
@pipermerriam I found this repetitive but I was missing an easy option to build my own `asynccontextmanager` using trio. So what I wanted was something like:

```python
async with run_geth() as ipc_path:
    async with run_dopple():
        # make requests and assertions
```

But not having an `@asynccontextmanager` I didn't feel like hand crafting these using a class and implementing some interface. Am I missing a special trio pattern do deal with these things?

Reply:
Why not just install `async-generator` and use it's version of `asynccontextmanager`?

Reply:
Wooot, that's awesome! I didn't know that exists!

Reply:
Ok, strange. I don't know where I failed before but it turns out that `contextlib.asynccontextmanager` does work out of the box with trio. So false alarm and I was able to greatly refactor and reduce the footprint of the whole setup.

Reply:
Anyway, I'm going to use this as a fallback to keep Python 3.6 support. I also added Python 3.6 specific integration test jobs.